src/content/contribute/writing-a-plugin.md

```diff
@@ -182,40 +183,92 @@ module.exports = HelloAsyncPlugin;
+Let's write a simple example plugin that generates a new build file called `assets.md`, the contents of which will list all of the asset files in our build. This plugin might look something like this:
 
 ```javascript
-class FileListPlugin {
+export class FileListPlugin {
+
+  static defaultOptions = {
+    outputFile: 'assets.md',
+  };
+
+  // Any options should be passed in the constructor of your plugin,
+  // (this is a public API of your plugin)
+  constructor(options = {}) {
+    // Applying user-specified options over the default options
+    // and making merged options further available to the plugin methods.
+    // You should probably validate all the options here as well
+    this.options = { ...FileListPlugin.defaultOptions, options };
+  }
+
   apply(compiler) {
-    // emit is asynchronous hook, tapping into it using tapAsync, you can use tapPromise/tap(synchronous) as well
-    compiler.hooks.emit.tapAsync('FileListPlugin', (compilation, callback) => {
-      // Create a header string for the generated file:
-      var filelist = 'In this build:\n\n';
-
-      // Loop through all compiled assets,
-      // adding a new line item for each filename.
-      for (var filename in compilation.assets) {
-        filelist += '- ' + filename + '\n';
-      }
 
-      // Insert this list into the webpack build as a new file asset:
-      compilation.assets['filelist.md'] = {
-        source: function () {
-          return filelist;
-        },
-        size: function () {
-          return filelist.length;
-        },
-      };
-
-      callback();
+    // webpack module instance can be accessed from the compiler object,
+    // this ensures that correct version of the module is used
+    const { webpack } = compiler;
+
+    // RawSource is one of the "sources" classes that should be used
+    // to represent asset sources in compilation
+    const { RawSource } = webpack.sources;
+
+    // Tapping to the "emit" hook
+    compiler.hooks.emit.tap(FileListPlugin.name, (compilation) => {
+
+      // "compilation.assets" is an object that contains all assets
+      // in the compilation, the keys of the object are pathnames of the assets
+      // and the values are file sources.
+
+      // Iterating over all the assets and
+      // generating content for our Markdown file.
+      const content = (
+        '# In this build:\n\n' +
+        Object.keys(compilation.assets)
+          .map((filename) => `- ${filename}`)
+          .join('\n')
+      );
+
+      // Adding new asset to the compilation, so it would be automatically
+      // generated by the webpack in the output directory
+      compilation.emitAsset(this.options.outputFile, new RawSource(content));
+
     });
   }
 }
+```
+
+```javascript
+import { FileListPlugin } from './file-list-plugin.js';
+
+// Use the plugin in your webpack configuration:
+export default {
```

Comment:
```suggestion
```javascript
const { FileListPlugin } = require('./file-list-plugin.js');

// Use the plugin in your webpack configuration:
module.exports = {
```
I would suggest using CommonJS in this page. We have to use the latest `webpack-cli` with either `webpack.config.mjs` or `type: "module"` to enable ESM support in webpack configuration, I think most users are not aware of this, and we're likely to confuse them in this page with those ESM syntaxes.